### PR TITLE
Add support for Basic Auth

### PR DESCRIPTION
**Pull request by gauravsarma1992**
https://github.com/DiceDB/dice/issues/141

Prior to Redis 6, authentication for a connection was enabled by passing in `requirepass` in the redis.conf or while starting the redis server. If the above config is set, prior to any command, the authentication has to be by running the following command:

```
AUTH '<password>'
```

[
Screenshot 2024-08-03 at 10 21 05 PM
](url)

`requirepass` can be set in the config by changing `config/main.go` before running the server.

The storage of user credentials and ACLs will be implemented in the subsequent PRs.

### CODE REVIEW

**Comment by JyotinderSingh** [comment]
Hi @gauravsarma1992, thanks for this contribution. Could you please rebase this PR on master and resolve the conflicts? Will review once done.

**Comment by gauravsarma1992** [comment]
> Hi @gauravsarma1992, thanks for this contribution. Could you please rebase this PR on master and resolve the conflicts? Will review once done.

Rebased the PR from the latest master.

**Comment by JyotinderSingh** [comment]
Please mark the PR ready for review once you are done with it.

**Comment by gauravsarma1992** [comment]
@JyotinderSingh Please review

File: core/session.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		}
+		if err = session.Activate(user); err != nil {
+			log.Println("error activating session for user", username, err)
+			continue
+		}
+		// User has been validated and the session has been activated
+		return
+	}
+	err = fmt.Errorf("no password matched for user %s", username)
+	return
```
Comment (arpitbbhayani) [inline core/session.go:148]: You are constructing the error and not returning it.

File: core/session.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	AuthCmd         = "AUTH"
+
+	SessionStatusPending SessionStatusT = SessionStatusT(0)
+	SessionStatusActive  SessionStatusT = SessionStatusT(1)
+	SessionStatusExpired SessionStatusT = SessionStatusT(2)
+)
+
+var (
+	UserStore *Users = NewUsers()
+)
```
Comment (arpitbbhayani) [inline core/session.go:24]: Redis must be storing the users in some durable storage, or is everything transient?

File: core/session.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	User struct {
+		Username          string
+		Passwords         []string
+		IsPasswordEnabled bool
+	}
+)
+
+func NewUsers() (users *Users) {
```
Comment (arpitbbhayani) [inline core/session.go:50]: Call this function `NewUserStore`

File: core/session.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	Users struct {
+		store  map[string]*User
+		stLock *sync.RWMutex
+	}
+
+	User struct {
+		Username          string
+		Passwords         []string
```
Comment (arpitbbhayani) [inline core/session.go:45]: Why multiple password for one user?

File: core/session.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		return
+	}
+	err = fmt.Errorf("no password matched for user %s", username)
+	return
+}
+
+func (session *Session) Expire() (err error) {
+	session.Status = SessionStatusExpired
+	return
+}
```
Comment (arpitbbhayani) [inline core/session.go:154]: When will the session expire? who will invoke this function and when?

File: core/session.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	return
+}
+
+func (users *Users) Get(username string) (user *User, err error) {
+	users.stLock.RLock()
+	defer users.stLock.RUnlock()
+	isPresent := false
+	if user, isPresent = users.store[username]; !isPresent {
+		err = fmt.Errorf("user %s not found", username)
```
Comment (arpitbbhayani) [inline core/session.go:63]: Reformat this error to how Redis raises it.

File: core/eval.go [hunk 7]
```diff
@@ -53,6 +54,38 @@ func evalPING(args []string) []byte {
...
 }
 
+// evalAUTH returns with an encoded "OK" if the user is authenticated
+// If the user is not authenticated, it returns with an encoded error message
+func evalAUTH(args []string, c *Client) []byte {
+	var (
+		b   []byte
+		err error
+	)
+	log.Print("Authenticating user")
```
Comment (arpitbbhayani) [inline core/eval.go:64]: Do not add log statements. We are yet to define logging standards.

File: core/eval.go [hunk 8]
```diff
@@ -1013,6 +1049,15 @@ func EvalAndRespond(cmds RedisCmds, c *Client) {
 	buf := bytes.NewBuffer(response)
 
 	for _, cmd := range cmds {
+		// Check if the command has been authenticated
+		if cmd.Cmd != AuthCmd && !c.Session.IsActive() {
+			// TODO: Check required response
+			log.Println("Session not active for CMD:", cmd.Cmd)
```
Comment (arpitbbhayani) [inline core/eval.go:1055]: remove log statements.

File: core/eval.go [hunk 9]
```diff
@@ -1013,6 +1049,15 @@ func EvalAndRespond(cmds RedisCmds, c *Client) {
...
+			if _, err := c.Write(RESP_AUTH_FAILURE); err != nil {
+				log.Panic(err)
```
Comment (arpitbbhayani) [inline core/eval.go:1057]: panic is not a good idea. Panic will terminate the server, which is bad.

**Review by arpitbbhayani — CHANGES_REQUESTED** [review]

File: core/eval.go (hunk 7 above)
Comment (gauravsarma1992) [reply core/eval.go]: Removed.

File: core/eval.go (hunk 8 above)
Comment (gauravsarma1992) [reply core/eval.go]: Removed.

File: core/eval.go (hunk 9 above)
Comment (gauravsarma1992) [reply core/eval.go]: Yes, oversight on my part. Removed the panic and logging the failure now.

File: core/session.go (hunk 2 above)
Comment (gauravsarma1992) [reply core/session.go]: Prior to Redis 6.0, there was only a concept of a static password which can be defined by passing `requirepass` in the configuration. This PR supports only that.
From Redis 6.0, the concept of ACLs was introduced. User auth storage along with finer grained permissions will be introduced in subsequent PRs.

File: core/session.go (hunk 4 above)
Comment (gauravsarma1992) [reply core/session.go:45]: Redis allows multiple passwords per single error. It compares all the passwords and returns authentication success if one of the set passwords match.

File: core/session.go (hunk 3 above)
Comment (gauravsarma1992) [reply core/session.go]: Done.

File: core/session.go (hunk 5 above)
Comment (gauravsarma1992) [reply core/session.go:154]: I wasn't able to find session expiry commands for Redis. Planning to keep this as introduction on ACLs may require this.

File: core/session.go (hunk 1 above)
Comment (gauravsarma1992) [reply core/session.go:148]: I am using a named return variable. `err` is defined in the function signature and will be returned automatically when calling `return`.

File: core/session.go (hunk 6 above)
Comment (gauravsarma1992) [reply core/session.go:63]: Given that this is an internal error which will not be returned to the client, does this require changing?

**Comment by gauravsarma1992** [comment]
@arpitbbhayani Addressed the review comments. Please review again.

**Comment by gauravsarma1992** [comment]
@JyotinderSingh Please review

**Review by JyotinderSingh — APPROVED** [review]

**Comment by arpitbbhayani** [comment]
Changes look good. Thanks :) merging them.

**Pull request merged**